New Integration: SMLIGHT SLZB-06 Adapters Integration

tl-sl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
SMLIGHT make a range of USB and LAN Zigbee coordinators. The LAN coordinators are based on ESP32 and have an onboard web UI. This integration aims to bring support for monitoring and managing these devices directly into Home Assistant. Providing direct access to sensors, switches, buttons and other features that are available from the SLZB web interface. There is no actual Zigbee support within this integration as that is handled per normal with either ZHA or Zigbee2MQTT.


Planned features:
* Management of device features (mode switch, restart, disable led's etc)
* Update sensors for both core firmware and zigbee firmware on SLZB-06 device
* Full coverage of various sensors available on device
* Configuration of language and timezones
* Firmware management for USB devices (SLZB-07)

This initial PR contains only a very slimmed done set of features for minimal usability for intial PR:
* Manual user config flow
* Zeroconf config flow
* Sensors platform including temp sensors, memory usage stats.

All variants of SLZB-06x are supported:
https://smlight.tech/product/slzb-06/#

Currently this integration requires SLZB-06 to have the v2.3.6.dev core firmware or later installed, a later PR will add support for slzb-06 with factory or older core firmware versions to allow the user to directly upgrade to a version  that fully supports this integration.

This integration is supported by SMLIGHT. code by @darkxst

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33053
https://github.com/home-assistant/brands/pull/5530
## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

tl-sl committed:
Initial SMLIGHT integration

Signed-off-by: Tim Lunn <tl@smlight.tech>

tl-sl committed:
Generated content

Signed-off-by: Tim Lunn <tl@smlight.tech>

ALERTua commented:
Hello. I had some time to test your new SMLight integration on my SLZB-06p10.
The integration was set up via Zeroconf.
![image](https://github.com/home-assistant/core/assets/8375129/038751b0-2bbb-4516-86a4-59a39b6d69a2)
Please tell me if you need more tests.
Thank you for your work! 💙💛

zweckj commented on homeassistant/components/smlight/coordinator.py:
```diff
@@ -0,0 +1,106 @@
+
+from .const import DOMAIN, LOGGER, SCAN_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
```
logger is defined and imported already use that

tl-sl replied:
yeh I cleaned up inconsistencies with LOGGER

zweckj commented on homeassistant/components/smlight/entity.py:
```diff
@@ -0,0 +1,39 @@
+from .const import ATTR_MANUFACTURER, DOMAIN
+from .coordinator import SmDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
```
idem

zweckj commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,166 @@
+
+UPTIME_DEVIATION = 5  # seconds
+
+_LOGGER = logging.getLogger(__name__)
```
idem

zweckj commented on homeassistant/components/smlight/__init__.py:
```diff
@@ -0,0 +1,47 @@
+from .const import DOMAIN
+from .coordinator import SmDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
```
idem

zweckj commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,166 @@
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=1,
+        value_fn=lambda x: x.sensors.esp32_temp,
```
pass in `x.sensors` to the `value_fn`

zweckj commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,166 @@
+
+SENSORS = [
+    SmSensorEntityDescription(
+        key="esp32_temperature",
```
is there a reason why the first two `key` and `translation_keys` are different? Might lead to confusion

tl-sl replied:
translation keys got renamed to user friendly names, but must have missed changing keys.

zweckj commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,166 @@
+        device_class=SensorDeviceClass.DATA_SIZE,
+        native_unit_of_measurement=UnitOfInformation.KILOBYTES,
+        entity_registry_enabled_default=False,
+        entity_category=EntityCategory.DIAGNOSTIC,
```
are you planning for non diagnostic sensors in the future? otherwise could move this to the description definitions

tl-sl replied:
1. I think all sensors will be diagnsotics will move

zweckj commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,166 @@
+    """Set up SMLIGHT sensor based on a config entry."""
+    coordinator: SmDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    sensors = [SmSensorEntity(coordinator, description) for description in SENSORS]
+    async_add_entities(sensors)
```
```suggestion
    async_add_entities(SmSensorEntity(coordinator, description) for description in SENSORS)
```

zweckj commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,166 @@
+    ) -> None:
+        """Initiate slzb sensor."""
+        super().__init__(coordinator)
+        CoordinatorEntity.__init__(self, coordinator)
```
base class

zweckj commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,166 @@
+        self._attr_unique_id = f"{coordinator.unique_id}_{description.key}"
+        self._last_uptime: datetime | None = None
+
+    def get_device_uptime(self, uptime: float | None) -> datetime:
```
I would move the whole uptime stuff to a separate sensor class that is independent from the standard sensors, would make the code clearer to read imo

tl-sl replied:
yes agree that is probably clearer

zweckj commented on homeassistant/components/smlight/strings.json:
```diff
@@ -0,0 +1,68 @@
+        "name": "Core SoC Temp"
+      },
+      "fs_usage": {
+        "name": "Filesystem Usage"
```
```suggestion
        "name": "Filesystem usage"
```

zweckj commented on homeassistant/components/smlight/strings.json:
```diff
@@ -0,0 +1,68 @@
+        "name": "Filesystem Usage"
+      },
+      "ram_usage": {
+        "name": "RAM Usage"
```
```suggestion
        "name": "RAM usage"
```
use sentence case (also for rest)

zweckj reviewed (commented):
some first feedback

zweckj commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,425 @@
+            {},
+        )
+
+        assert result2["type"] == FlowResultType.FORM
```
compare those with `is`

zweckj commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,425 @@
+            data=DISCOVERY_INFO,
+        )
+        assert result["type"] == FlowResultType.FORM
+        assert result.get("step_id") == "confirm_discovery"
```
no `.get` necessary on those

zweckj commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,425 @@
+            },
+        )
+
+        assert result2["type"] == FlowResultType.FORM
```
always finish the tests to terminal states, to show we can recover (create entry/abort) by fixing the error and continuing

tl-sl replied:
I have fixed this one to recover with correct credentials, will see which others it makes sense for, some of the errors are really fatal though.

zweckj commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,425 @@
+) -> None:
+    """Test we abort zeroconf flow if device already configured."""
+    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
+    with (
```
I believe we'll need a fixture here to avoid all those individual patches

tl-sl replied:
yes I agree, should have done that from the start, will implement

zweckj commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,425 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result.get("type") == FlowResultType.FORM
```
access directly `[]`

zweckj commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,425 @@
+        )
+
+        assert result3.get("type") is FlowResultType.CREATE_ENTRY
+        assert result3 == snapshot
```
I don't really like snapshots in config_flow tests. Imo it's better to test the data directly

tl-sl replied:
ok, have dropped snapshot from here.

zweckj commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,425 @@
+    assert result.get("type") is FlowResultType.FORM
+    assert result.get("step_id") == "confirm_discovery"
+
+    mock_info = Info.from_dict(loads(load_fixture("smlight/info.json")))
```
probably better in a fixutre as well

darkxst committed:
Cleanup LOGGING

darkxst committed:
Call super first

darkxst committed:
coordinator instance attributes

darkxst committed:
Move coordinatorEntity and attr to base class

tl-sl commented:
Thanks for the review Zweckj, I will work these items shortly ;)

darkxst committed:
cleanup sensors

darkxst committed:
Use runtime data

darkxst committed:
update strings to use sentence case

darkxst committed:
Use fixture for config_flow tests and test to completion

darkxst committed:
Split uptime hndling into a new uptime sensor entity

tl-sl commented on homeassistant/components/smlight/__init__.py:
```diff
@@ -0,0 +1,47 @@
+    await coordinator.async_config_entry_first_refresh()
+    coordinator.unique_id = format_mac(coordinator.data.info.MAC).replace(":", "")
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```
ok, sure

tl-sl commented on homeassistant/components/smlight/config_flow.py:
```diff
@@ -0,0 +1,202 @@
+class SmlightConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for SMLIGHT Zigbee."""
+
+    VERSION = 1
```
done

tl-sl commented on homeassistant/components/smlight/entity.py:
```diff
@@ -0,0 +1,39 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class SmEntity:
```
Will end up with two different Coordinators (one for local device data, and another for internet based data such as available firmware updates), but can probablty inherit here and handle update sensors differently when they come.

darkxst committed:
Add tests for init, coordinator and sensor

darkxst committed:
Drop server side events and internet callback

will bring this back with binary sensor Platform

darkxst committed:
consolidate coordinator setup

darkxst committed:
Improve reauth flow on incorrect credentials

tl-sl commented:
Thanks for the review @zweckj, I have now  addressed all your suggestions. Along with various other minor clean ups/improvements and added some additional test coverage for integration init and sensor platform.

zweckj commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,167 @@
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
```
Type with your type

zweckj commented on homeassistant/components/smlight/__init__.py:
```diff
@@ -0,0 +1,29 @@
+]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
You are missing a custom type for your entry like `type SmConfigEntry = ConfigEntry[SmDataUpdateCoordinator]`

darkxst committed:
Use custom type SmConfigEntry

tl-sl commented on tests/components/smlight/test_sensor.py:
```diff
@@ -0,0 +1,60 @@
+    """Test platform setup."""
+
+    await setup_platform(hass, mock_config_entry, SENSOR_DOMAIN)
+
```
Initially wasnt sure how to handle uptime sensors in a snapshot, but found it.

darkxst committed:
entity always include connections

darkxst committed:
get_hostname tweak

tl-sl commented:
Thanks again, I have resolved these remaining items now also. Also added a few improvement to test fixtures and a few more tests to get 100% coverage.

tl-sl committed:
update sensor snapshot

gam-st commented:
@zweckj are there any options to get this reviewed and merged?

ALERTua commented:
A test from a random user (me):
![image](https://github.com/user-attachments/assets/e0618d4b-68d3-4faa-91c3-45f190c2830c)

joostlek commented on homeassistant/components/smlight/coordinator.py:
```diff
@@ -0,0 +1,92 @@
+            return SmData(
+                sensors=await self.client.get_sensors(),
+                info=await self.client.get_info(),
+                internet=False,
```
Why is this always false

tl-sl replied:
I removed the code that handles the internet sensor for a later PR. this was somehow left over will remove.

joostlek commented on homeassistant/components/smlight/entity.py:
```diff
@@ -0,0 +1,33 @@
+            manufacturer=ATTR_MANUFACTURER,
+            model=coordinator.data.info.model,
+            name=coordinator.hostname,
+            sw_version=f"core: {coordinator.data.info.sw_version} / zb: {coordinator.data.info.zb_version}",
```
What's a zb?

joostlek commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,175 @@
+        return self.entity_description.value_fn(self.coordinator.data.sensors)
+
+
+class SmUptimeSensorEntity(SmSensorEntity):
```
Why do we have a separate class for this?

tl-sl replied:
I split out the uptime sensors into a separate class based on suggestions from @zweckj in an earlier review. Mainly to separate the uptime handling/logic from plain sensors.

However I can revert this back to a single class similar to how I had it originally, if that would be preferred?

This was earlier revision of my sensor class:
https://github.com/home-assistant/core/blob/3403d235a98fb823da13aa9d19ce362baa811b0e/homeassistant/components/smlight/sensor.py#L109

Let me know if its preferred to merge it back into a single class and I will update accordingly.

darkxst replied:
I have merged uptime sensors back in the main sensor class now.

joostlek replied:
Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen

sergge1 replied:
> Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen

The review was postponed until changes were submitted. However, this PR is good with both single and separate classes, aligning with project practices. Can this be processed further with a single class? Or should it be separated?

gam-st replied:
> Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen

The review was postponed until changes were submitted. However, this PR is good with both single and separate classes, aligning with project practices. Can this be processed further with a single class? Or should it be separated?

gam-st replied:
> Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen

Because 
- The review was postponed until changes were submitted. 
- That was a part of "joostlek **requested changes**  [4 days ago]". 
Answer to the question is [here](https://github.com/home-assistant/core/pull/118675#discussion_r1698400920).

Just to clarify, this PR is good to go with both single and separate classes, aligning with project practices. 
Should it be separated again for your approval? Just let PR author know what is needed, i believe reverting back to separate classes is just a few minutes task.

gam-st replied:
> Note: this wasn't an action to do something, rather an exploratory question to know why this is chosen
my 3rd version of this reply :)

changes were made because that was a part of "joostlek requested changes [4 days ago]". - misunderstanding probably. And, it blocks further commit checks as the status is still "changes requested" 
The answer to your question is here https://github.com/home-assistant/core/pull/118675#discussion_r1698400920.
Just to clarify, this PR is good to go with both single and separate classes, aligning with project practices.
Should it be separated again for your approval? Just please let the PR author know what is needed, i believe reverting back to separate classes is just a few minutes task.

joostlek replied:
No it was more that the code looked like a lot, so instead of trying to understand what you are trying to do I wanted to consult what the code is doing so I then can use that info to check the code

gam-st replied:
> No it was more that the code looked like a lot, so instead of trying to understand what you are trying to do I wanted to consult what the code is doing so I then can use that info to check the code

Coear

> No it was more that the code looked like a lot, so instead of trying to understand what you are trying to do I wanted to consult what the code is doing so I then can use that info to check the code

Got it, clear now, thanks, Joostlek. 
So to sum up, regarding the single class or separate classes question: 
what should be done:
1. to keep it as a single class (it is already as single), or 
2. to make it through multiple classes? (both options are valid and works). 

May it remain "as is" in the current PR and  users get it merged? Can you please review this  PR?

tl-sl replied:
> No it was more that the code looked like a lot, so instead of trying to understand what you are trying to do I wanted to consult what the code is doing so I then can use that info to check the code

Splitting the sensor class does add some complexity in itself. I not sure this outweighs the resulting simplification in the native_value property. Thus I could go either way on this. Regardless I have tried to clarify further the uptime handling code.

joostlek replied:
But then my questions still stands, what extra calculation do we have to do for the uptime sensor?

tl-sl replied:
> But then my questions still stands, what extra calculation do we have to do for the uptime sensor?

There is no extra calculations for uptime just what you see there now.

The uptime is reported from the device in "Seconds since last boot", this is converted to a fixed datetime stamp, to minimise state changes of the sensor.  The deviation buffer allows any changes caused by clock skew etc to be ignored (we dont want to update the sensor state just because the calculated datetime varies by a few milliseconds). The HA frontend takes care of converting the uptime into a user friendly "3 hours" uptime.

This is very much specific of implementation of the sensor in HA, and the calculation also requires state only known within HA. So I think it makes more sense to have here, rather than in the library.

darkxst committed:
Drop reauth flow for later PR

darkxst committed:
Use _async_setup for initial setup

darkxst committed:
drop internet to be set later

darkxst committed:
sensor fixes

darkxst committed:
typing fixes

darkxst committed:
Bump pysmlight dependency to 0.0.12

darkxst committed:
config flow re

tl-sl reviewed (commented):
Thanks for the review @joostlek.

I have addressed most of your comments. See inline responses.  One question remaining in there regarding the uptime sensor class, I split this class out due to feedback in earlier review. If preferred I can revert this back to just including the uptime sensors in the main sensor class.

darkxst committed:
dont trigger invalid auth message when first loading auth step

ALERTua commented:
If the PR is ready for review, maybe it should be un-drafted? ;)
My tests are still successful with this integration and my SLZB-06P10.

tl-sl commented:
Thanks for testing @ALERTua 

> If the PR is ready for review, maybe it should be un-drafted? ;)

Just waiting on a response to this question https://github.com/home-assistant/core/pull/118675#discussion_r1698400920, then I will mark it ready to review again.

gam-st commented:
@zweckj Hello Josef, 
@joostlek Hello Joost,
any options to get this reviewed?

darkxst committed:
Merge uptime sensors back into main sensor class

darkxst committed:
clarify uptime handling

alesoft73 commented:
We await the new integration....... ;)

KoKolaj commented:
It looks very interesting.

gam-st commented:
@joostlek 
@zweckj 
are there any options to get this reviewed and merged now when all comments addressed and answers provided?

joostlek commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,163 @@
+class SmSensorEntityDescription(SensorEntityDescription):
+    """Class describing SMLIGHT sensor entities."""
+
+    entity_category: EntityCategory = EntityCategory.DIAGNOSTIC
```
```suggestion
    entity_category = EntityCategory.DIAGNOSTIC
```

joostlek commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,163 @@
+        entity_registry_enabled_default=False,
+        value_fn=lambda x: x.fs_used,
+    ),
+    SmSensorEntityDescription(
+        key="core_uptime",
+        translation_key="core_uptime",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        entity_registry_enabled_default=False,
+        value_fn=lambda x: x.uptime,
+    ),
+    SmSensorEntityDescription(
+        key="socket_uptime",
+        translation_key="socket_uptime",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        entity_registry_enabled_default=False,
+        value_fn=lambda x: x.socket_uptime if x.socket_uptime > 0 else None,
```
What do theser eturn?

tl-sl replied:
socket uptime is reported as 0 when it is disconnected, but we want this to display as `Unavailable`. I probably should move this to the library. removed now, will fix for uptime PR

joostlek commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,163 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up SMLIGHT sensor based on a config entry."""
+    coordinator: SmDataUpdateCoordinator = entry.runtime_data
```
```suggestion
    coordinator = entry.runtime_data
```

joostlek commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,163 @@
+
+
+class SmSensorEntity(SmEntity, SensorEntity):
+    """Representation of a slzb sensor."""
+
+    def __init__(
+        self,
+        coordinator: SmDataUpdateCoordinator,
+        description: SmSensorEntityDescription,
+    ) -> None:
+        """Initiate slzb sensor."""
+        super().__init__(coordinator)
+
+        self.entity_description: SmSensorEntityDescription = description
```
```suggestion
    """Representation of a slzb sensor."""
    
    entity_description: SmSensorEntityDescription

    def __init__(
        self,
        coordinator: SmDataUpdateCoordinator,
        description: SmSensorEntityDescription,
    ) -> None:
        """Initiate slzb sensor."""
        super().__init__(coordinator)

        self.entity_description = description
```

joostlek commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,163 @@
+        self._attr_unique_id = f"{coordinator.unique_id}_{description.key}"
+        self._last_uptime: datetime | None = None
+
+    def get_uptime(self, uptime: float | None) -> datetime | None:
+        """Return device or zigbee socket uptime.
+
+        Converts uptime from seconds to a datetime value, allow up to 5
+        seconds deviation. Fire a reboot event if device uptime has reset
+        since last run.
+        """
+        if uptime is None:
+            # reset to unknown state
+            self._last_uptime = None
+            return None
+
+        delta = timedelta(seconds=uptime)
+
+        if "core_uptime" in self.entity_description.key and delta <= SCAN_INTERVAL:
+            self.fire_reboot_event()
+
+        new_uptime = utcnow() - delta
+
+        if (
+            not self._last_uptime
+            or abs(new_uptime - self._last_uptime) > UPTIME_DEVIATION
+        ):
+            self._last_uptime = new_uptime
+
+        return self._last_uptime
```
I think I agree that this could be a simple class overwriting the native_value function to handle this. Ideally I think we can better move this to a separate PR to allow the PR to be merged quicker

tl-sl replied:
OK, have removed uptime sensors for a seperate PR.

joostlek commented on homeassistant/components/smlight/sensor.py:
```diff
@@ -0,0 +1,163 @@
+
+        return self._last_uptime
+
+    def fire_reboot_event(self):
+        """Fire reboot event."""
+        self.coordinator.hass.bus.async_fire(
+            SMLIGHT_SLZB_REBOOT_EVENT,
+            {
+                "device_id": self.coordinator.unique_id,
+                "host": self.coordinator.hostname,
+            },
+        )
+        LOGGER.debug("SLZB device reboot detected")
```
I'd rather see an event entity displaying this

tl-sl replied:
sure, will do in uptime PR

joostlek commented on homeassistant/components/smlight/strings.json:
```diff
@@ -0,0 +1,58 @@
+  },
+  "entity": {
+    "sensor": {
+      "zigbee_temperature": {
+        "name": "Zigbee SoC temp"
+      },
+      "core_temperature": {
+        "name": "Core SoC temp"
+      },
```
SoC?

tl-sl replied:
shorthand for "System On a Chip", but perhaps clearer if just say "Zigbee chip temp"?

joostlek commented on homeassistant/components/smlight/strings.json:
```diff
@@ -0,0 +1,58 @@
+      "socket_uptime": {
+        "name": "Zigbee uptime"
+      },
+      "wifi_status": {
+        "name": "Wi-Fi status"
+      }
```
unused

tl-sl replied:
removed

tl-sl committed:
address review comments

tl-sl committed:
Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

tl-sl committed:
pass host as parameter to the dataCoordinator

tl-sl committed:
update sensor test snapshot

tl-sl committed:
drop uptime sensors for a later PR

tl-sl committed:
move coordinator unique_id to _async_setup

tl-sl commented on homeassistant/components/smlight/__init__.py:
```diff
@@ -0,0 +1,32 @@
+    """Set up SMLIGHT Zigbee from a config entry."""
+    coordinator = SmDataUpdateCoordinator(hass, entry)
+    await coordinator.async_config_entry_first_refresh()
+    coordinator.unique_id = format_mac(coordinator.data.info.MAC)
```
This needed data from after the first refresh, however I have now moved it to _async_setup and explicitly get info data from api client.

tl-sl commented on homeassistant/components/smlight/config_flow.py:
```diff
@@ -0,0 +1,150 @@
+
+        if user_input is not None:
+            self.host = user_input[CONF_HOST]
+            assert self.host is not None
```
this was because mypy complaining about type of `self.host` being optional, but param to Api2 is mandatory. I changed to use an intermediate host variable instead, hopefully that is clearer?

tl-sl commented on homeassistant/components/smlight/config_flow.py:
```diff
@@ -0,0 +1,150 @@
+    ) -> ConfigFlowResult:
+        """Handle authentication to SLZB-06 device."""
+        errors: dict[str, str] = {}
+        self.context["confirm_only"] = False
```
yes this is, I thought I needed to clear it from an earlier discovery step, but in fact not required. removed.

tl-sl commented on homeassistant/components/smlight/coordinator.py:
```diff
@@ -0,0 +1,82 @@
+        super().__init__(
+            hass,
+            LOGGER,
+            name=f"{DOMAIN}_{entry.data[CONF_HOST]}",
```
ok

tl-sl reviewed (commented):
thanks @joostlek, I have addressed these comments now.

Have removed uptime sensors and associated code for separate PR.

darkxst committed:
fix CI

joostlek commented on tests/components/smlight/test_init.py:
```diff
@@ -0,0 +1,42 @@
+    """Test async_setup_entry."""
+    assert mock_config_entry.state is ConfigEntryState.LOADED
+    assert mock_config_entry.unique_id == "aa:bb:cc:dd:ee:ff"
+    assert isinstance(mock_config_entry.runtime_data, SmDataUpdateCoordinator)
```
idem

joostlek commented on tests/components/smlight/test_sensor.py:
```diff
@@ -0,0 +1,62 @@
+
+
+@pytest.mark.usefixtures("entity_registry_enabled_by_default")
+@pytest.mark.parametrize(
+    "entity_id",
+    [
+        "sensor.slzb_06_core_chip_temp",
+        "sensor.slzb_06_zigbee_chip_temp",
+        "sensor.slzb_06_ram_usage",
+        "sensor.slzb_06_filesystem_usage",
+    ],
+)
+async def test_sensors(
+    hass: HomeAssistant,
+    device_registry: dr.DeviceRegistry,
+    entity_registry: er.EntityRegistry,
+    snapshot: SnapshotAssertion,
+    entity_id: str,
+) -> None:
+    """Test the SMLIGHT sensors."""
+
+    assert (state := hass.states.get(entity_id))
+    assert state == snapshot
+
+    assert (entry := entity_registry.async_get(entity_id))
+    assert entry == snapshot
+
+    assert entry.device_id
+    assert (device_entry := device_registry.async_get(entry.device_id))
+    assert device_entry == snapshot
```
Please use the `snapshot_platform` helper instead

tl-sl replied:
ok done

joostlek commented on tests/components/smlight/test_sensor.py:
```diff
@@ -0,0 +1,62 @@
+    assert (entry := entity_registry.async_get(entity_id))
+    assert entry == snapshot
+
+    assert entry.device_id
+    assert (device_entry := device_registry.async_get(entry.device_id))
+    assert device_entry == snapshot
```
Snapshot the device in a test_init.py test

tl-sl replied:
moved device snapshot to test_init.py

joostlek commented on tests/components/smlight/test_sensor.py:
```diff
@@ -0,0 +1,62 @@
+    assert device_entry == snapshot
+
+
+@pytest.mark.parametrize(
+    "entity_id",
+    [
+        "sensor.slzb_06_ram_usage",
+        "sensor.slzb_06_filesystem_usage",
+    ],
+)
```
Rather use a for loop than parametrize here

tl-sl committed:
Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

tl-sl committed:
drop invalid_auth test tag

tl-sl committed:
use snapshot_platform, update fixtures

tl-sl commented on tests/components/smlight/test_coordinator.py:
```diff
@@ -0,0 +1,59 @@
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """Test coordinator get hostname from IP."""
+    coordinator = mock_config_entry.runtime_data
```
@joostlek I not entirely clear what to do here instead? Should I create a new instance of the Coordinator just for the test? Or perhaps test these at a higher level as part of config_flow or async_setup?

joostlek replied:
Neither. The coordinator is just a tool to receive data, we consider that as internals so I would strongly discourage not to use a test_coordinator.py. It all depends on what you're testing, because it will fetch data, and it will display state using that data. So with what you're testing here, where does the user notice this?

tl-sl replied:
ok got it. Come to think of it, I can probably drop the hostname handling entirely, it was only used for a name in the device registry. I will also remove test_coordinator.

tl-sl committed:
Finish all tests with abort or create entry

tl-sl commented on tests/components/smlight/conftest.py:
```diff
@@ -0,0 +1,94 @@
+            loads(load_fixture("sensors.json", DOMAIN))
+        )
+
+        if request.node.get_closest_marker("invalid_auth"):
+            api.check_auth_needed.return_value = True
+            api.authenticate.side_effect = SmlightAuthError
+        else:
+            api.check_auth_needed.return_value = False
+            api.authenticate.return_value = True
```
ok

tl-sl commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,365 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result2["type"] is FlowResultType.CREATE_ENTRY
+    assert result2["context"]["source"] == "user"
+    assert result2["title"] == "SLZB-06p7"
+    assert result2["data"] == {
+        CONF_HOST: MOCK_HOST,
+    }
```
sure

tl-sl commented on tests/components/smlight/test_config_flow.py:
```diff
@@ -0,0 +1,365 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result2["type"] is FlowResultType.FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
+    assert result2["step_id"] == "confirm_discovery"
```
ok, this is done now.

tl-sl committed:
drop coordinator tests and remove hostname support

tl-sl committed:
add test for update failure on connection error

tl-sl commented on tests/components/smlight/test_coordinator.py:
```diff
@@ -0,0 +1,59 @@
+) -> None:
+    """Test coordinator update failed."""
+    mock_smlight_client.get_info.side_effect = SmlightConnectionError
+    coordinator = mock_config_entry.runtime_data
```
will test update failures a different way

joostlek commented on tests/components/smlight/test_init.py:
```diff
@@ -0,0 +1,92 @@
+
+    mock_smlight_client.get_info.side_effect = SmlightConnectionError
+
+    next_update = dt_util.utcnow() + timedelta(minutes=10)
+    async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
```
Please use the freezer

tl-sl committed:
use freezer for update_failed test

tl-sl committed:
fix pysmlight imports

tl-sl commented:
Thanks for your reviews on this Joostlek and Zweckj. I will now get to work on some followup PR's ;)
